docstrings, mostly {readers,writers}.py

razofz opened the pull request:
More docstrings. Mostly files readers.py and writers.py. Need some input for a few things. 

- `writers.py`
  - function `subset_assay_zarr` in writers.py has a really messy docstring right now. To use the function one seems to need knowledge about numpy and zarr indexing, which leaves the question: how much of that should we explain? Should we just link to their docs?
  - all `dump` methods, maybe you could write a description there @parashardhapola ?
  - `dask_to_zarr`, correct that param `df` is a dask array? or a dataframe?
- `readers.py`
  - `CrReader`, description for param `grp_names`
  - what does the `consume*` methods do?
  - type of param `a` for function `to_sparse` in `MtxDirReader`?
  - `H5adReader`, description for param `groupCodes`

razofz commented:
update: added some more. Went through the vignettes and checked that everything used therein was documented (since what's used there is most likely to be used by our users, at least initially). Made some grammatical changes to vignette texts while I was there.

```
Overall statistics for 19 files (skipped all non-init magic methods) (skipped private methods):
Needed: 282  -  Found: 203  -  Missing: 79
Total coverage: 72.0%  -  Grade: Very good 
```

parashardhapola commented on scarf/readers.py:
```diff
@@ -34,28 +41,38 @@ def get_file_handle(fn: str) -> IO:
         else:
             return open(fn, 'r')
     except (OSError, IOError, FileNotFoundError):
-        raise FileNotFoundError("ERROR: FILE NOT FOUND: %s" % fn)
```
Shouldn't this be `fn`

razofz replied:
oh yes, this must have just been an accidental deletion of a character, sorry

parashardhapola commented on scarf/assay.py:
```diff
@@ -57,10 +56,10 @@ class Assay:
         nthreads:
         rawData:
-        feats:
+        feats: a dataframe with info about each feature in the dataset
```
Technically `feats` is not a dataframe. It is a Metadata object.

razofz replied:
True! I realised that for attr cells in other places, will change this here as well

parashardhapola commented on scarf/readers.py:
```diff
@@ -311,6 +411,13 @@ def _read_dataset(self, key: Optional[str] = None):
+        Returns the input data as a sparse (COO) matrix.
+
+        Args:
+            a:
```
`a` here is a dense numpy matrix.

razofz replied:
thanks!

razofz commented:
do you have any input on the other questions I asked above?

parashardhapola commented:
Looks great. Will merge after one more scan through the diff. 
Is there a badge to track doc coverage?
I would be nice to have it on main README.md

Now the important thing would be to see how these are parsed by sphinx and displayed on the API section of readthedocs.

Next steps would be to improve the toctrees in the sphinx docs.

Thanks

parashardhapola commented:
> More docstrings. Mostly files readers.py and writers.py. Need some input for a few things.
> 
> * `writers.py`
>   
>   * function `subset_assay_zarr` in writers.py has a really messy docstring right now. To use the function one seems to need knowledge about numpy and zarr indexing, which leaves the question: how much of that should we explain? Should we just link to their docs?

**Good question:  The idea here was to expose this functionality for power users. But most users will use DataSore.make_subset() which wraps this.**


>   * all `dump` methods, maybe you could write a description there @parashardhapola ?

**Yes, I can take care of that.
I suggest we create a new issue about missing docstrings and create a task list there. I would also help me keep track of whats missing and resolve it. Can you create a new issue and create a starter list there?**

>   * `dask_to_zarr`, correct that param `df` is a dask array? or a dataframe?
> * `readers.py`

**This is simply poor choice of variable naming on my part.  `df` is infact  a dask array.
We should not change this at this stage (to avoid breaking changes) but create a TODO there maybe. I can then add a deprecation warning there and change it in near future.**

>   * `CrReader`, description for param `grp_names`

**grp_names is the dictionary that tells CrReader where to find the matrix, features and barcodes. Since CrReader is an abstract class I do not expect users to directly use it.** 

>   * what does the `consume*` methods do?

**`consume` methods  are generators that yield chunks of data from external file formats like MTX, H5ad, Loom.  The chunks of data are in sparse matrix format and contain `n` number of cells. Each chunk contains all the features for the `n` cells.  A uniform `consume` API across all reader classes ensures a smooth interoperability with writer methods that actually consume `consume`** 

>   * type of param `a` for function `to_sparse` in `MtxDirReader`?

**`a` is  a sparse matrix but as numpy.ndarray  type. It contains a chunk of data from MTX file. The `consume` method that calls `to_sparse` makes sure that cells are not spilled across chunks.**

>   * `H5adReader`, description for param `groupCodes`

`groupCodes` attribute is designed to deal with two ways in which the data is stored in H5ad. Older versions of Anndata used ot store data as in `var` and `obs` as  `h5py.Datasets` (from H5py library). In these datasets all the cell attributes like thier barcode ids, clusters, etc were saved in a matrix where each column is an attribute. But later versions of anndata started using a different approach by storing each column as separate dataset. This caused `var` and `obs` to become `h5py.Groups` . `groupCodes` is designed to keep track of this.

These some very good questions. Can you see if answers to some of them can be added as comments in the code?



``

razofz commented:
> >   * function `subset_assay_zarr` in writers.py has a really messy docstring right now. To use the function one seems to need knowledge about numpy and zarr indexing, which leaves the question: how much of that should we explain? Should we just link to their docs?
> 
> **Good question: The idea here was to expose this functionality for power users. But most users will use DataSore.make_subset() which wraps this.**

Allright, I'm with you. Maybe I'll cut down that docstring and provide a link to the numpy/zarr docs then. 

> > * all `dump` methods, maybe you could write a description there @parashardhapola ?
> 
> **Yes, I can take care of that. I suggest we create a new issue about missing docstrings and create a task list there. I would also help me keep track of whats missing and resolve it. Can you create a new issue and create a starter list there?**

Sure, that sounds like a good idea! I'll get on that.

> These some very good questions. Can you see if answers to some of them can be added as comments in the code?

Thanks for the answers! I'll try to incorporate them into the docstrings.

razofz commented:
updated docstrings according to discussion. 

Regarding the `consume*` methods, they seem like mostly internal methods even if they aren't marked as private (with underscore(s) as prefix), so I took a minimal approach in documenting them. Didn't add docstrings to the subclasses of `CrReader`, but they import the description from their superclass when one calls `help()` on them. Only downside with this is that they won't count towards a docstring coverage calculation.

> I suggest we create a new issue about missing docstrings and create a task list there. I would also help me keep track of whats missing and resolve it. Can you create a new issue and create a starter list there?

Added issue #33 as task list.

parashardhapola commented on scarf/writers.py:
```diff
@@ -583,7 +558,7 @@ def subset_assay_zarr(zarr_fn: str, in_grp: str, out_grp: str,
 
 def dask_to_zarr(df, z, loc, chunk_size, nthreads: int, msg: str = None):
-    # TODO: docstring, correct that param df is a dask array?
+    # TODO: perhaps change name of Dask array so it does not get confused with a dataframe
```
good idea!
